File: Week4/VolvoCodingChallenge/app/src/main/java/com/yeah/ruisu/volvocodingchallenge/data/remote/RemoteServiceHelper.java
package com.yeah.ruisu.volvocodingchallenge.data.remote;


import android.provider.SyncStateContract;

import com.yeah.ruisu.volvocodingchallenge.data.remote.models.CityID;
import com.yeah.ruisu.volvocodingchallenge.data.remote.models.ForecastDatum;
import com.yeah.ruisu.volvocodingchallenge.utils.Constants;

import java.util.List;

import io.reactivex.Single;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class RemoteServiceHelper
{
    public Single<CityID> getWOEID(String cityName)
    {
        Retrofit MyRetroFit = getRetroFit(Constants.URLS.WEATHER_API_BASE_URL);
        RemoteService MyService = MyRetroFit.create(RemoteService.class);

        return MyService.getWOEID(cityName);
    }

    public Single<List<ForecastDatum>> getWeatherForecast(String WOEID,
                                                          String year,
                                                          String month,
                                                          String day)
    {
        Retrofit MyRetroFit = getRetroFit(Constants.URLS.WEATHER_API_BASE_URL);
        RemoteService MyService = MyRetroFit.create(RemoteService.class);

        return MyService.getWeatherForecast(WOEID, year, month, day);
    }

    private Retrofit getRetroFit(String baseUrl)
    {
        HttpLoggingInterceptor MyInterceptor = new HttpLoggingInterceptor();

        MyInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient MyClient = new OkHttpClient.Builder()
                                                .addInterceptor(MyInterceptor)
                                                .build();

        return new Retrofit.Builder()
                            .baseUrl(baseUrl)
                            .client(MyClient)
                            .addConverterFactory(GsonConverterFactory.create())
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .build();
    }
}

File: Reference/DeltaChallenge-master/DeltaChallenge-master/app/src/main/java/com/example/admin/deltachallenge/data/remote/RemoteServiceHelper.kt
package com.example.admin.deltachallenge.data.remote

import com.example.admin.deltachallenge.data.models.NumbersResponse
import com.example.admin.deltachallenge.utils.Constants
import com.jakewharton.retrofit2.adapter.kotlin.coroutines.experimental.CoroutineCallAdapterFactory
import kotlinx.coroutines.experimental.Deferred
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class RemoteServiceHelper {

    fun getRandomNumbers(): Deferred<NumbersResponse> {
        val retrofit = getRetrofit(Constants.BASE_URL)
        val service = retrofit.create(RemoteService::class.java)
        return service.getRandomNumbers()
    }

    private fun getRetrofit(baseUrl: String): Retrofit {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(interceptor).build()

        return Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(CoroutineCallAdapterFactory())
                .build()
    }
}